hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocation.java

```diff
@@ -176,28 +177,34 @@ public void notifyException(Throwable exception) {
             return;
         }
 
-        if ((isBindToSingleConnection() && exception instanceof IOException)
-                || System.currentTimeMillis() > retryTimeoutPointInMillis) {
+        if (isBindToSingleConnection() && exception instanceof IOException) {
             clientInvocationFuture.complete(exception);
             return;
         }
 
-        if (isRetrySafeException(exception)
+        boolean isAllowedToRetry = isRetrySafeException(exception)
                 || invocationService.isRedoOperation()
-                || (exception instanceof TargetDisconnectedException && clientMessage.isRetryable())) {
-            try {
-                ClientExecutionServiceImpl executionServiceImpl = (ClientExecutionServiceImpl) this.executionService;
-                executionServiceImpl.schedule(this, RETRY_WAIT_TIME_IN_SECONDS, TimeUnit.SECONDS);
-            } catch (RejectedExecutionException e) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Retry could not be scheduled ", e);
-                }
-                clientInvocationFuture.complete(exception);
+                || (exception instanceof TargetDisconnectedException && clientMessage.isRetryable());
+        if (!isAllowedToRetry) {
+            clientInvocationFuture.complete(exception);
+            return;
+        }
+
+        long msAfterTimeout = System.currentTimeMillis() - retryTimeoutPointInMillis;
+        if (msAfterTimeout > 0) {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Exception will not be retried because invocation timed out", exception);
             }
+            clientInvocationFuture.complete(new OperationTimeoutException("Invocation timed out by " + msAfterTimeout + " ms"));
```

Comment:
Probably it would be good to capture more information. So when has this call started, address, partition id etc etc etc. On the serverside there is a lot of information available when the infamous OperationTimeoutException is thrown.